azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+    stage.pipeline_root = stage
+    callback_called = threading.Event()
+
+    def callback(*arg, **argv):
+        assert threading.current_thread().name == "callback"
+        callback_called.set()
+
+    stage.on_pipeline_event_handler = callback
+
+    stage.handle_pipeline_event(event)
+    callback_called.wait()
+
+
 TestPipelineRootStagePipelineThreading.test_runs_callback_in_callback_thread = (
     _test_pipeline_root_runs_callback_in_callback_thread
 )
 TestPipelineRootStagePipelineThreading.test_runs_operation_in_pipeline_thread = (
     _test_pipeline_root_runs_operation_in_pipeline_thread
 )
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_connected_in_callback_thread = (
+    _test_pipeline_root_runs_on_connected_in_callback_thread
+)
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_disconnected_in_callback_thread = (
+    _test_pipeline_root_runs_on_disconnected_in_callback_thread
+)
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_event_received_in_callback_thread = (
+    _test_pipeline_root_runs_on_event_received_in_callback_thread
+)
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    cls=pipeline_stages_base.EnsureConnectionStage,
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=[
+        pipeline_ops_mqtt.MQTTPublishOperation,
+        pipeline_ops_mqtt.MQTTSubscribeOperation,
+        pipeline_ops_mqtt.MQTTUnsubscribeOperation,
+    ],
+    all_events=all_common_events,
+    handled_events=[],
+)
+
+fake_topic = "__fake_topic__"
+fake_payload = "__fake_payload__"
+ops_that_cause_connection = [
+    {
+        "op_class": pipeline_ops_mqtt.MQTTPublishOperation,
+        "op_init_kwargs": {"topic": fake_topic, "payload": fake_payload},
+    },
+    {"op_class": pipeline_ops_mqtt.MQTTSubscribeOperation, "op_init_kwargs": {"topic": fake_topic}},
+    {
+        "op_class": pipeline_ops_mqtt.MQTTUnsubscribeOperation,
+        "op_init_kwargs": {"topic": fake_topic},
+    },
+]
+
+
+@pytest.mark.parametrize(
+    "params",
+    ops_that_cause_connection,
+    ids=[x["op_class"].__name__ for x in ops_that_cause_connection],
+)
+@pytest.mark.describe("EnsureConnectionStage - .run_op() -- called with connect operation")
+class TestEnsureConnectionStageRunOp(object):
+    @pytest.fixture
+    def op(self, mocker, params):
+        op = params["op_class"](**params["op_init_kwargs"])
+        op.callback = mocker.MagicMock()
+        return op
+
+    @pytest.fixture
+    def stage(self, mocker):
+        stage = make_mock_stage(
+            mocker=mocker, stage_to_make=pipeline_stages_base.EnsureConnectionStage
+        )
+        stage.next.run_op = mocker.MagicMock()
+        return stage
+
+    @pytest.mark.it("Passes the operation down the pipline when the transport is already connected")
+    def test_operation_alrady_connected(self, params, op, stage):
```

Comment:
This test class is kind of tricky, because technically this method has four units/paths:

1) Called with an op that needs connection, and we are not connected
2) Called with an op that needs connection, and we are connected
3) Called with an op that does not need connection, and we are not connected
4) Called with an op that does not need connection, and we are connected

In practice, units 2-4 are all the exact same path, which only requires a single test, so for simplicity you've thrown that single test in with all the others from unit 1. This can be a little confusing while reading the output, but honestly, it's not the end of the world, since we DO have the coverage.

Reply:
Added in TFS task 5117988

---
In reply to: [312667014](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312667014) [](ancestors = 312667014)